File: ported/auto/4122a3.js
// Auto-translated from Ghidra C by tools/c-to-js/translate.js.
// Source: decompiled/c/4122a3.c
// Edit by hand only after diff-test passes — re-running the translator will overwrite.

/** @typedef {import("../../runtime/heap.js").Heap} Heap */

import { mmioAdvance, mmioGetInfo, mmioSetInfo } from "../../runtime/win32.js";
import { regs } from "../../runtime/regs.js";
import { FUN_004138d0 } from "./4138d0.js";
export function FUN_004122a3(heap, param_1, param_2, param_3, param_4, param_5) {
  const __sp = heap.allocFrame(128);
  const __addr_local_5c = __sp + 0;
  const __addr_local_14 = __sp + 72;
  const __addr_local_10 = __sp + 76;
  const __addr_local_c = __sp + 80;
  const __addr_local_8 = __sp + 84;
  try {
  let MVar1 = 0;
  heap.setU32(__addr_local_8, (0) >>> 0);
  MVar1 = ((mmioGetInfo(heap, param_1, __addr_local_5c, 0)) >>> 0);
  heap.setU32(__addr_local_8, ((((MVar1 != 0)) | 0)) >>> 0);
  if (heap.u32(__addr_local_8) == 0) {
    heap.setU32(__addr_local_14, (param_2) >>> 0);
    if (heap.u32((param_4 + 4)) < param_2) {
      heap.setU32(__addr_local_14, (heap.u32((param_4 + 4))) >>> 0);
    }
    heap.setI32((param_4 + 4), (heap.i32((param_4 + 4)) - heap.u32(__addr_local_14)) & 0xffffffff);
    for (heap.setU32(__addr_local_c, (0) >>> 0); heap.u32(__addr_local_c) < heap.u32(__addr_local_14); heap.setU32(__addr_local_c, (heap.u32(__addr_local_c) + heap.u32(__addr_local_10)) >>> 0)) {
      if (heap.u32((__addr_local_5c + 16)) == heap.u32((__addr_local_5c + 28))) {
        heap.setU32(__addr_local_8, (mmioAdvance(heap, param_1, __addr_local_5c, 0)) >>> 0);
        if (heap.u32(__addr_local_8) != 0) {
          heap.setU32(param_5, (0) & 0xffffffff);
          return heap.u32(__addr_local_8);
        }
        heap.setU32(__addr_local_8, (0) >>> 0);
        if (heap.u32((__addr_local_5c + 16)) == heap.u32((__addr_local_5c + 28))) {
          heap.setU32(__addr_local_8, (0xe103) >>> 0);
          heap.setU32(param_5, (0) & 0xffffffff);
          return heap.u32(__addr_local_8);
        }
      }
      heap.setU32(__addr_local_10, (((heap.u32((__addr_local_5c + 28))) | 0) - ((heap.u32((__addr_local_5c + 16))) | 0)) >>> 0);
      if (heap.u32(__addr_local_14) - heap.u32(__addr_local_c) < heap.u32(__addr_local_10)) {
        heap.setU32(__addr_local_10, (heap.u32(__addr_local_14) - heap.u32(__addr_local_c)) >>> 0);
      }
      (regs.eax = FUN_004138d0(heap, heap.u32(__addr_local_c) + param_3, heap.u32((__addr_local_5c + 16)), heap.u32(__addr_local_10)));
      heap.setU32((__addr_local_5c + 16), (heap.u32((__addr_local_5c + 16)) + heap.u32(__addr_local_10)) >>> 0);
    }
    heap.setU32(__addr_local_8, (mmioSetInfo(heap, param_1, __addr_local_5c, 0)) >>> 0);
    if (heap.u32(__addr_local_8) == 0) {
      heap.setU32(param_5, (heap.u32(__addr_local_14)) & 0xffffffff);
      return 0;
    }
  }
  LAB_004123e9: heap.setU32(param_5, (0) & 0xffffffff);
  return heap.u32(__addr_local_8);
} finally {
    heap.freeFrame(128);
  }
}

File: decompiled/c/4122a3.c

MMRESULT FUN_004122a3(HMMIO param_1,uint param_2,int param_3,int param_4,uint *param_5)

{
  MMRESULT MVar1;
  _MMIOINFO local_5c;
  uint local_14;
  uint local_10;
  uint local_c;
  MMRESULT local_8;
  
  local_8 = 0;
  MVar1 = mmioGetInfo(param_1,&local_5c,0);
  local_8 = (MMRESULT)(MVar1 != 0);
  if (local_8 == 0) {
    local_14 = param_2;
    if (*(uint *)(param_4 + 4) < param_2) {
      local_14 = *(uint *)(param_4 + 4);
    }
    *(int *)(param_4 + 4) = *(int *)(param_4 + 4) - local_14;
    for (local_c = 0; local_c < local_14; local_c = local_c + local_10) {
      if (local_5c.pchNext == local_5c.pchEndRead) {
        local_8 = mmioAdvance(param_1,&local_5c,0);
        if (local_8 != 0) goto LAB_004123e9;
        local_8 = 0;
        if (local_5c.pchNext == local_5c.pchEndRead) {
          local_8 = 0xe103;
          goto LAB_004123e9;
        }
      }
      local_10 = (int)local_5c.pchEndRead - (int)local_5c.pchNext;
      if (local_14 - local_c < local_10) {
        local_10 = local_14 - local_c;
      }
      FUN_004138d0(local_c + param_3,local_5c.pchNext,local_10);
      local_5c.pchNext = local_5c.pchNext + local_10;
    }
    local_8 = mmioSetInfo(param_1,&local_5c,0);
    if (local_8 == 0) {
      *param_5 = local_14;
      return 0;
    }
  }
LAB_004123e9:
  *param_5 = 0;
  return local_8;
}

